feat: allow user to provide custom templates for project initialization

lorenzo-cavazzi opened the pull request:
Add the possibility to select a custom source for templates.
The user can insert a repository URL and reference (tag, commit or branch) and try to fetch the templates. If the repo is a valid Renku templates repo, all the available templates will be listed in the Templates dropdown, otherwise a (hopefully not too cryptic) error will be shown.
If needed, we can try to map the most common errors to something more user-friendly, but I assume everyone trying to create custom template repositories would have pretty solid technical skills.

![Screenshot_20200810_181351](https://user-images.githubusercontent.com/43481553/89805026-4a07db80-db35-11ea-9ace-5c0709bb0c18.png)

PREVIEW: https://lorenzotest.dev.renku.ch/projects/new

fix #976

vfried commented:
I found a little bug, I added a project, then was redirected to the project page, then went again to add a project and got this error... Looks like the visibility wasn't reinitialized

![image](https://user-images.githubusercontent.com/42647877/89984256-61e87800-dc79-11ea-8843-e2a0eca41c91.png)

vfried commented on src/project/new/ProjectNew.present.js:
```diff
@@ -438,52 +430,84 @@ class TemplateSource extends Component {
+            invalid={this.state.missingRef} />
+          <FormFeedback>Provide a template repository reference.</FormFeedback>
+          <FormText>
+            <FontAwesomeIcon icon={faInfoCircle} /> Preferably a tag or a commit. A branch is also valid,
```
It would be really cool if we could fetch the commits from the repo so they can choose from a dropdown, but I guess that's not that important for now...

lorenzo-cavazzi replied:
I would rather leave a one-liner here, I believe a user playing with custom templates knows about fixed vs dynamic git references. If this turns out to be confusing, we can change this later.
If you find a more explanatory short sentence please feel free to suggest it, I'm happy to change it.

lorenzo-cavazzi replied:
> It would be really cool if we could fetch the commits from the repo so they can choose from a dropdown, but I guess that's not that important for now...

Ah yeah, that would make a better user experience! It's not a super quick change, I will open another issue for this :+1:

vfried replied:
I think we can leave it like this and if users don't understand we can just add something there. I also think fetching the templates in another issue is a good idea, is not something we need to have now.

lorenzo-cavazzi commented:
Thanks for thoroughly testing the project creation!

I forgot to merge master before deploying :facepalm:  The namespace bug seems gone, please let me know if it's not the case.
I fixed a bug in the validation function that caused the false warning on the visibility and I added a piece of code to reset the input data after a project has been successfully created. I still leave the templates, I think there is no need to re-fetch them (the RenkuLab templates won't change and re-entering the custom templates repository data may be an annoyance the user don't want to repeat).

Preview updated here: https://lorenzotest.dev.renku.ch/projects/new

I see your point about the consistency with the form generator. I went for a different implementation because I used a couple of non-standard components combined with realtime validation, but I guess I can improve the layout to have a similar appearance. Should we move this discussion to another issue since it targets the whole "New project" page and it's a small re-factory?
